Leave typecasts on variables on stack

mgudemann opened the pull request:
Instead of removing all typecasts from variables on the stack, we extract the
necessary information of the bare expression from a temporary variable via a
`reference_wrapper`.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f7602f80979ee50c733f9b5a3d84848a264f5481).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77881179

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3184,28 +3184,38 @@ void java_bytecode_convert_methodt::save_stack_entries(
+    if(
+      (write_type == bytecode_write_typet::VARIABLE ||
+       write_type == bytecode_write_typet::STATIC_FIELD) &&
+      entry.get().id() == ID_symbol)
```
It would be nice if the formatting cleanup could be factored out into a separate commit.

mgudemann replied:
unfortunately, `clang-format` seems to work on the whole condition, I applied it only on the second conjunct, same for the other cases

tautschnig replied:
Oh, I'm sorry, I missed the `stack_entry` -> `entry` change.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3184,28 +3184,38 @@ void java_bytecode_convert_methodt::save_stack_entries(
+    while(entry.get().id() == ID_typecast)
+    {
+      entry = std::ref(to_typecast_expr(stack_entry).op());
+    }
```
There already is code doing `skip_typecast`, but that's just local. I have created #2508 to change that.

tautschnig replied:
#2508 has now been merged, hence this could be rewritten to `const exprt &entry = skip_typecast(stack_entry);`

tautschnig commented:
I am not sure why the target of this PR is "java-stack-typecasts" instead of "develop"?

mgudemann commented:
@tautschnig please see https://github.com/diffblue/cbmc/pull/2498
This is an alternative to solve the problem at hand. Merging this th the branchand then squashing into #2498 would be the way to go. Fine with your approach to typecast unwrappring, will update once that's merged. I'd also like to have a test for the issue in #2498

mgudemann commented:
@tautschnig rebased onto develop, used your suggestion

Commit:
Keep expressions unchanged when adding temporary variables

Search recursively through expressions instead of only removing outer typecast
expressions.

mgudemann commented:
@tautschnig Updated with @kroening s suggestion in #2498

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -3173,44 +3174,87 @@ irep_idt java_bytecode_convert_methodt::get_static_field(
 {
+  const std::function<bool(
+    const std::function<tvt(const exprt &expr)>, const exprt &expr)>
+    entry_matches = [&entry_matches, &identifier](
```
Perhaps easier to use `depth_begin()` et al and just continue applying your predicate until it gives a definitive answer?

mgudemann replied:
not familiar with `depth_begin` I guess it iterates over the whole `exprt` tree?

smowton reviewed (commented):
Two major comments:
1. Is this changing behaviour / fixing a bug? If so, add tests.
2. Add to function documentation to indicate why all these different cases exist -- what kind of temp variable does it make for a member, for a deref, for a symbol etc?

mgudemann commented:
@smowton yes, there's a bug where TG crashes, I will reproduce the bytecode sequence in Jasmin and add a test. In short the code before removed type casts and that gave rise to a typecheck error in some (at least one) cases.

Commit:
Simplify stack element replacement loop

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 93e0ab4f9a0f746db1fbd3f56a17489b057cbdb3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78913856
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

Commit:
Add regression test for stack variables with typecasts

This regression test mimicks the bytecode that exhibited the error in Tika.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a63212e1385bce778ae24ccf58f6964276b29b5a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79554978

peterschrammel commented on jbmc/regression/jbmc/stack_var12/test.desc:
```diff
@@ -0,0 +1,17 @@
+
+ buffer[position++];
+
+The code pushes position, duplicates it and the modifies the stack
```
then modifies?
